### ISSUE TITLE
`fnm upgrade` command

### ISSUE THREAD

**Issue by waldyrious** [Schniz/fnm#781]
It would be nice to have an `fnm upgrade` command that would upgrade the current Node version to the latest minor version, and optionally remove the old ole (e.g. with an extra CLI option, like `--rm-old`).

Alternatively, if upgrading minors is considered too dangerous to be the default, this command could upgrade just the patch version, and upgrading the minor version would be done by explicitly passing a separate CLI option (e.g. `fnm upgrade --minor`).

**Comment by satvikpendem** [comment]
I actually would like it to upgrade the major version and then for minor or patches we could have flags.

**Comment by waldyrious** [comment]
@satvikpendem there's already `fnm install --lts` / `fnm install lts-latest` to install the latest LTS. [Apparently](https://github.com/Schniz/fnm/issues/644#issuecomment-1018448684) there's no convenient shorthand to install the latest non-LTS version, but that functionality should IMO come in the form of an `fnm install latest` and/or `fnm install --latest`, rather than as part of an `upgrade` subcommand.

**Comment by alexeyten** [comment]
Isn't `fnm install` already behave as you want?
`fnm install 14` will install latest `14.x.x` (`14.20.0` at this time).

**Comment by waldyrious** [comment]
Yeah, but if I'm on 14 I'd like `fnm upgrade` to bump me to the latest 14.x.x, whereas if I'm on 16 I'd like the same command to bump me to the latest 16.x.x. Of course it's possible to do that in other ways, it's just a convenience shortcut that I feel `fnm` could offer easily, building upon its existing functionality.

**Comment by armujahid** [comment]
**Requirements:** 

1. `fnm upgrade` should upgrade to latest patch version and optionally to latest minor version using something like `fnm upgrade --minor`
2. it should migrate global packages from old version to new version

Old version can be manually removed later after verification and testing. Doing this in upgrade command is risky.
Currently for step 2 I am using this approach https://github.com/Schniz/fnm/issues/481#issuecomment-1292726042

Ideally, we need something like `--reinstall-packages-from` option in fnm
https://stackoverflow.com/a/34810922/2073920

In nvm, we can do `nvm install NEW_VERSION --reinstall-packages-from=OLD_VERSION`
**Other step 2 related issues:**
https://github.com/Schniz/fnm/issues/481
https://github.com/Schniz/fnm/issues/620

**Comment by libook** [comment]
> **Requirements:**
> 
>     1. `fnm upgrade` should upgrade to latest patch version and optionally to latest minor version using something like `fnm upgrade --minor`
> 
>     2. it should migrate global packages from old version to new version
> 
> 
> Old version can be manually removed later after verification and testing. Doing this in upgrade command is risky. Currently for step 2 I am using this approach [#481 (comment)](https://github.com/Schniz/fnm/issues/481#issuecomment-1292726042)
> 
> Ideally, we need something like `--reinstall-packages-from` option in fnm [stackoverflow.com/a/34810922/2073920](https://stackoverflow.com/a/34810922/2073920)
> 
> In nvm, we can do `nvm install NEW_VERSION --reinstall-packages-from=OLD_VERSION` **Other step 2 related issues:** #481 #620

I have submitted pull requests before.

https://github.com/Schniz/fnm/pull/510

Need a Shell-less implementation, such as using rust to call npm directly.

**Comment by nklowns** [comment]
Hi, I'd love to have this feature too.

**Comment by LitoMore** [comment]
I would like to have a a command to upgrade fnm itself as well.

I'm using a special platform which does have such application managers. I'm using the `curl -fsSL https://fnm.vercel.app/install | bash` for installing fnm.

I know I can rerun this command to upgrade this application, but I'd prefer to have a more convenient way for upgrading.

**Comment by jathek** [comment]
> I would like to have a command to upgrade fnm itself as well.

This should probably be a new issue.

**Comment by LitoMore** [comment]
> This should probably be a new issue. (@jathek)

I don't think so. The `upgrade` is better for update fnm itself. Like `bun upgrade`, `deno upgrade` does.

**Comment by jathek** [comment]
> > This should probably be a new issue. (@jathek)
> 
> I don't think so. The `upgrade` is better for update fnm itself. Like `bun upgrade`, `deno upgrade` does.

I think you should still open a separate issue for upgrading fnm itself, which is not a thing it currently does. This issue was opened to add a feature to improve Node versions updates. Maybe you think the `upgrade` argument should be reserved for your feature, but you've already said that here. You should just link the new issue here instead of trying to steer this one.

**Comment by SukkaW** [comment]
For now, I have been doing `fnm install 22`, then `fnm ls`, and then manually uninstalling all previous Node.js installations within the same major version. I want a way to automatic this. Things like `fnm upgrade 22` would be a nice to have.

**Comment by brandon-d-mckay** [comment]
Here's a PowerShell 7 function you can tweak to your needs:
```
function top::fnmu {
    $original = fnm current

    fnm use default
    $default = fnm current

    $latest = fnm list-remote --latest
    if($default -ne $latest){
        fnm install --latest
        fnm default $latest
        fnm use default
        fnm uninstall $default
    }

    if($original -ne $default){ fnm use $original }
}
```

**Comment by SukkaW** [comment]
Here is my current upgrade:

```sh
  function fnm() {
    if [[ $1 == "upgrade" ]]; then
        local matched_verions=()
        local pattern_match_version="* v$2"
        local pattern_match_system="* system $2"

        local is_system=0
        local line=""

        command fnm ls | while read LINE; do
            line=${LINE}

            if (( $line[(I)$pattern_match_version] )); then
                matched_verions+=${${${line#* v}%% *}%% default}
            fi

            if (( $line[(I)$pattern_match_system] )); then
                is_system=1
            fi
        done

        for version in $matched_verions; do
            echo "Uninstalling Node.js $version"
            command fnm uninstall $version
        done

        command fnm install "$2"

        if (( $is_system )); then
            command fnm alias system $2
        fi
    else
      command fnm "$@"
    fi
  }
```

Now you can do things like `fnm upgrade 18`. It will uninstall any version prefixed with `18`, then reinstall the latest 18 w/ `fnm install 18`. If you have previously done `fnm alias system 18`, this script can also help you re-alias.

**Comment by stevehipwell** [comment]
Coming from [nvs](https://github.com/jasongin/nvs) the only thing I miss is being able to run `nvs upgrade 24` to upgrade the major version in place.

**Comment by stevehipwell** [comment]
@SukkaW NVS is Windows only, so as much as I prefer Bash to PowerShell, your script isn't going to help me out for this use case.

**Comment by fin-major** [comment]
Would be great to have this implemented.

Could be similar to how Python Install Manager does it:
pymanager install [major-version] --update

For example:
pymanager install 3.14 --update
